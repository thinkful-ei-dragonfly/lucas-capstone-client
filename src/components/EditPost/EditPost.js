import React from 'react'
import PostApiService from '../../services/post-api-services'
import FileBase64 from 'react-file-base64';
import PostContext from '../../PostContext/PostContext'
import config from '../../config'



export default class EditPost extends React.Component {
  static contextType = PostContext || {}
  state = {
    post: {},
    uploadedFile: null,
    post_type: 'Text'
  }

  successfulSubmission = () => {
    const { location, history } = this.props
    const destination = (location.state || {}).from || '/'
    history.push(destination)
  }
  generateMessage(string) {
    return (
      <p>{string}</p>
    )
  }
  componentDidMount() {
    const postId = this.props.match.params.post_id
      return fetch(`${config.API_ENDPOINT}/posts/${postId}`)
        .then(res => {
          return res.json()
        })
        .then(res => {
          this.setState({
            post: res
          })
        })
    }
  handleSubmit = ev => {
    ev.preventDefault()
    ev.persist()
    const { post_type, title, text_title, text_content, caption, video, audio} = ev.target
    let newPost = {
      id: this.props.match.params.post_id,
      type: this.state.post_type,
      title: title.value
    }

    switch (post_type.value) {
      case 'Text':
        newPost.text_title = text_title.value
        newPost.text_content = text_content.value
        break;
      case 'Image':
        newPost.image = this.state.uploadedFile.base64
        newPost.caption = caption.value
        break;
      case 'Video':
        newPost.video = video.value
        newPost.caption = caption.value
        break
      case 'Audio':
        newPost.audio = audio.value
        newPost.caption = caption.value
        break;
      default:
    }
    PostApiService.updatePost(newPost)
      .then(res => {
        if (!res.ok) {
          this.generateMessage(res.error)
        }
        this.generateMessage('Your post was successfully updated')

        switch (this.state.post_type) {
          case 'Text':
            title.value = ''
            text_title.value = ''
            text_content.value = ''
            break;
          case 'Image':
            title.value = ''
            caption.value = ''
            break;
          case 'Video':
            title.value = ''
            video.value = ''
            break;
          case 'Audio':
            title.value = ''
            audio.value = ''
            caption.value = ''
            break;
          default:
        }
        setTimeout(() => {
          this.successfulSubmission()
        }, 500)

      })
  }

  updatePostType = e => {
    this.setState({
      post_type: e.target.value
    })

  }
  getFile(file){
    this.setState({ uploadedFile: file })
  }

  renderFieldTypes () {
    let fields = ''
    switch (this.state.post_type) {
      case 'Image':
        fields = (
          <>
            <label htmlFor='image'>Insert Image Address</label>

            <FileBase64
              id='image'
              name='image'
              accept='image/png, image/jpeg'
              multiple={ false }
              onDone={ this.getFile.bind(this) }
            />
            <label htmlFor='caption'>Image Caption</label>
            <textarea
              type='text'
              id='caption'
              name='caption'
              placeholder={this.state.post.caption}
              ></textarea>
          </>
        )
        break;
      case 'Text':
        fields = (
          <>
            <label htmlFor='text_title'>Headline</label>
            <input
              type='text'
              id='text_title'
              name='text_title'
              placeholder='Headline'
              defaultValue={this.state.post.text_title}
            >
          </input>
          <label htmlFor='text_content'>Body</label>
          <textarea
            type='text'
            id='text_content'
            name='text_content'
            placeholder={this.state.post.text_content}
            ></textarea>
          </>
        )
        break;
      case 'Video':
        fields = (
          <>
            <label htmlFor='video'>Enter a Vimeo ID below</label>
            <input
              type='text'
              id='video'
              name='video'
              defaultValue={this.state.post.video}
            >
            </input>
            <label htmlFor='caption'>Video Caption</label>
            <textarea
              type='text'
              id='caption'
              name='caption'
              placeholder={this.state.post.caption}
              ></textarea>
          </>
        )
        break;
      case 'Audio':
      fields = (
        <>
          <label htmlFor='audio'>Insert Audio File Location</label>

          <input
            id='audio'
            name='audio'
            accept='image/png, image/jpeg'
          />
          <label htmlFor='caption'>Audio Caption</label>
          <textarea
            type='text'
            id='caption'
            name='caption'
            ></textarea>
        </>
      )
        break;
      default:

    }
    return (
      <div className='add-post-fields'>
        {fields}
      </div>
    )

  }

  render() {
    return (
      <form
        className='AddPostForm'
        onSubmit={this.handleSubmit}
      >
      <div className='form-title'>
        <h2>Edit Post</h2>
      </div>
      <div className='message'>
        {this.generateMessage()}
      </div>
      <div className='select'>
        <label htmlFor='post_type'>Post Type</label>
        <select
          required
          aria-label="Select your post type"
          name='post_type'
          id='post_type'
          onChange={this.updatePostType}
          defaultValue={this.props.post_type}
        >
        <option value='Text'>Text</option>
        <option value='Image'>Image</option>
        <option value='Video'>Video</option>
        </select>

      </div>
      <div className='form-fields'>
        <label htmlFor='title'>Post Title</label>
        <input type='text' name='title' id='title' placeholder='Post Title' defaultValue={this.state.post.title}></input>
        {this.renderFieldTypes()}
      </div>
      <button
        type='submit'
        className='form-submit'
        >Submit</button>
      </form>
    )
  }
}
