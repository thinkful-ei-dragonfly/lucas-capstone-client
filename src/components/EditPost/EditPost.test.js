import React from 'react'
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json'

import EditPost from './EditPost'
configure({ adapter: new Adapter() });
describe('EditPost component', () => {
  const props = {
    match: {
      params: {
        post_id: "a2c7df25-b59a-4769-809a-9c9517c0e7ce"
      }
    }
  }

  it('renders a EditPost by default', () => {
    const wrapper = shallow(<EditPost {...props}/>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  it('renders a form', () => {
    const form = shallow(<EditPost {...props}/>)
    expect(toJson(form)).toMatchSnapshot()
  })
  it('renders a form-title label in the form', () => {
    const form_title = shallow(<EditPost {...props}/>)
      .find('.form-title')
    expect(toJson(form_title)).toMatchSnapshot()
  })
  it('renders a select label in the form', () => {
    const select = shallow(<EditPost {...props}/>)
      .find('.select')
    expect(toJson(select)).toMatchSnapshot()
  })
  it('renders a "post_type" dropdown menu', () => {
    const post_type = shallow(<EditPost {...props}/>)
      .find('#post_type')
    expect(toJson(post_type)).toMatchSnapshot()
  })
  it('renders form fields', () => {
    const form_fields = shallow(<EditPost {...props}/>)
      .find('.form-fields')
    expect(toJson(form_fields)).toMatchSnapshot()
  })
})
