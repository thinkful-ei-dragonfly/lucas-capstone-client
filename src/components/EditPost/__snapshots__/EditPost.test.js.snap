// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPost component renders a "post_type" dropdown menu 1`] = `
<select
  aria-label="Select your post type"
  id="post_type"
  name="post_type"
  onChange={[Function]}
  required={true}
>
  <option
    value="Text"
  >
    Text
  </option>
  <option
    value="Image"
  >
    Image
  </option>
  <option
    value="Video"
  >
    Video
  </option>
</select>
`;

exports[`EditPost component renders a EditPost by default 1`] = `
<form
  className="AddPostForm"
  onSubmit={[Function]}
>
  <div
    className="form-title"
  >
    <h2>
      Edit Post
    </h2>
  </div>
  <div
    className="message"
  >
    <p />
  </div>
  <div
    className="select"
  >
    <label
      htmlFor="post_type"
    >
      Post Type
    </label>
    <select
      aria-label="Select your post type"
      id="post_type"
      name="post_type"
      onChange={[Function]}
      required={true}
    >
      <option
        value="Text"
      >
        Text
      </option>
      <option
        value="Image"
      >
        Image
      </option>
      <option
        value="Video"
      >
        Video
      </option>
    </select>
  </div>
  <div
    className="form-fields"
  >
    <label
      htmlFor="title"
    >
      Post Title
    </label>
    <input
      id="title"
      name="title"
      placeholder="Post Title"
      type="text"
    />
    <div
      className="add-post-fields"
    >
      <label
        htmlFor="text_title"
      >
        Headline
      </label>
      <input
        aria-label="Text Headline"
        id="text_title"
        name="text_title"
        placeholder="Headline"
        type="text"
      />
      <label
        htmlFor="text_content"
      >
        Body
      </label>
      <textarea
        aria-label="Text Body"
        id="text_content"
        name="text_content"
        type="text"
      />
    </div>
  </div>
  <button
    aria-label="Submit Form"
    className="form-submit"
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`EditPost component renders a form 1`] = `
<form
  className="AddPostForm"
  onSubmit={[Function]}
>
  <div
    className="form-title"
  >
    <h2>
      Edit Post
    </h2>
  </div>
  <div
    className="message"
  >
    <p />
  </div>
  <div
    className="select"
  >
    <label
      htmlFor="post_type"
    >
      Post Type
    </label>
    <select
      aria-label="Select your post type"
      id="post_type"
      name="post_type"
      onChange={[Function]}
      required={true}
    >
      <option
        value="Text"
      >
        Text
      </option>
      <option
        value="Image"
      >
        Image
      </option>
      <option
        value="Video"
      >
        Video
      </option>
    </select>
  </div>
  <div
    className="form-fields"
  >
    <label
      htmlFor="title"
    >
      Post Title
    </label>
    <input
      id="title"
      name="title"
      placeholder="Post Title"
      type="text"
    />
    <div
      className="add-post-fields"
    >
      <label
        htmlFor="text_title"
      >
        Headline
      </label>
      <input
        aria-label="Text Headline"
        id="text_title"
        name="text_title"
        placeholder="Headline"
        type="text"
      />
      <label
        htmlFor="text_content"
      >
        Body
      </label>
      <textarea
        aria-label="Text Body"
        id="text_content"
        name="text_content"
        type="text"
      />
    </div>
  </div>
  <button
    aria-label="Submit Form"
    className="form-submit"
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`EditPost component renders a form-title label in the form 1`] = `
<div
  className="form-title"
>
  <h2>
    Edit Post
  </h2>
</div>
`;

exports[`EditPost component renders a select label in the form 1`] = `
<div
  className="select"
>
  <label
    htmlFor="post_type"
  >
    Post Type
  </label>
  <select
    aria-label="Select your post type"
    id="post_type"
    name="post_type"
    onChange={[Function]}
    required={true}
  >
    <option
      value="Text"
    >
      Text
    </option>
    <option
      value="Image"
    >
      Image
    </option>
    <option
      value="Video"
    >
      Video
    </option>
  </select>
</div>
`;

exports[`EditPost component renders form fields 1`] = `
<div
  className="form-fields"
>
  <label
    htmlFor="title"
  >
    Post Title
  </label>
  <input
    id="title"
    name="title"
    placeholder="Post Title"
    type="text"
  />
  <div
    className="add-post-fields"
  >
    <label
      htmlFor="text_title"
    >
      Headline
    </label>
    <input
      aria-label="Text Headline"
      id="text_title"
      name="text_title"
      placeholder="Headline"
      type="text"
    />
    <label
      htmlFor="text_content"
    >
      Body
    </label>
    <textarea
      aria-label="Text Body"
      id="text_content"
      name="text_content"
      type="text"
    />
  </div>
</div>
`;
